### ISSUE TITLE
Smart layouts, ways to avoid reverse engineering.

### ISSUE THREAD

**Issue by scoiatael** [opendesigndev/octopus-specs#25]
In GitLab by @nikitamironov on Mar 16, 2020, 14:53

### [hardcore-full.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/8a7ee6b68dfc09599a3a1c3557a444d13efdfac6812c99724cfe846a070b6ea5/hardcore-full.sketch)
Nejvyssi komplexita, trva asi 4 minuty: 16 artboardu s 100 smart layout instancemi v kazdem. Kazda instance ma 3 urovne layoutu.
```
time sketchtool detach --output=h.sketch ./hardcore.sketch 

real	3m50.566s
user	3m17.710s
sys	0m10.035s
```
### [hardcore-8a.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/ed65f1005d2b8bf3c341ece185f6f1805f5d546d06c7747a74d510630209d4ea/hardcore-8a.sketch)
Same, ale 8 artboardu misto 16.
```
time sketchtool detach --output=h.sketch ./hardcore-8a.sketch 

real	1m33.686s
user	1m32.168s
sys	0m3.689s
```
### [hardcore-4a.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/1b8f8ef7cb0252d2df853347c51a44c48c44d71a787aeac6202cc1c131b0a488/hardcore-4a.sketch)
Same, ale 4 artboardu misto 16.
```
time sketchtool detach --output=h.sketch ./hardcore-4a.sketch 

real	0m45.574s
user	0m45.008s
sys	0m1.529s
```
### [hardcore-2a.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/93a22bd8cdee55f32a0291cdd36b56fe23ceca62b4fa0c436c1de2b32038cc28/hardcore-2a.sketch)
Same, ale 2 artboardu misto 16.
```
time sketchtool detach --output=h.sketch ./hardcore-2a.sketch 

real	0m48.779s
user	0m47.261s
sys	0m1.722s
```
### [hardcore-1a.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/1bbc4af8b4a92a8f026b8a2e4fa72e5c36dcfc12fb63f183b5bfb33ae54502c2/hardcore-1a.sketch)
Same, ale 1 artboard misto 16.
```
time sketchtool detach --output=h.sketch ./hardcore-1a.sketch 

real	0m11.728s
user	0m11.409s
sys	0m0.458s
```
### [hardcore-1a-half.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/7b1e7f2d9c7ac403011063587f28ae47002dc36a5ed3d0d12838e6dccdc4d832/hardcore-1a-half.sketch)
```
time sketchtool detach --output=h.sketch ./hardcore-1a-half.sketch 

real	0m5.985s
user	0m5.887s
sys	0m0.219s
```
### [hardcore-1a-quart.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/4e7c27529d6a2877664b0f8b58cfa42d456786e88be008d8dd053283d0ce078f/hardcore-1a-quart.sketch)
```
time sketchtool detach --output=h.sketch ./hardcore-1a-quart.sketch 

real	0m2.893s
user	0m2.880s
sys	0m0.131s
```
### [hardcore-1a-three.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/ee3dfd9e8345f2d4cf99d913209ce76ba56887c3584988300280a77215131bdb/hardcore-1a-three.sketch)
```
time sketchtool detach --output=h.sketch ./hardcore-1a-three.sketch 

real	0m1.749s
user	0m1.543s
sys	0m0.112s
```
### [hardcore-full-reduced-complexity.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/13900f261f93618385b7d1cbd503207012254541e2c943c786cc0edcb4a5a81a/hardcore-full-reduced-complexity.sketch)
Cely design (vsichni artboardy), ale detachnuta jedna uroven layoutu (pred celym detachem).
```
time sketchtool detach --output=h.sketch ./hardcore-full-reduced-complexity.sketch 

real	1m33.397s
user	1m34.134s
sys	0m4.407s
```
### [hardcore-full-reduced-complexity-3.sketch](https://my-gitlab-bucket.s3.amazonaws.com/569732949/55f09c91cd7dd513b0930c5a051ded273a17c42640a2a865fe1214b132d0597f/hardcore-full-reduced-complexity-3.sketch)
Cely design (vsichni artboardy), ale detachnute dve urovne layoutu (pred celym detachem).
```
time sketchtool detach --output=h.sketch ./hardcore-full-reduced-complexity-3.sketch 

real	0m46.004s
user	0m44.815s
sys	0m1.498s
```

Z toho vyplyva, ze pro designy aktivne vyuzivajici smart layouty to urcite nebude rychly, orientacne to bude trvat 1-2 minuty jenom pro detach. Zkouseli jsme s @jakubzitny i jine realne designy, vetsinou jejich specifika je to, ze soubory jsou mnohem vetsi (pro srovnani tohle je 1.5mb, kdyz jsme zkouseli designy zakazniku na 200-300mb testovacich souborech), ale presto jsou rychlejsi pro detach - ono to tolik nezalezi na rozmeru sobouru kolik na jeho algoritmicke komplexite. Ty velke soubory pro srovnani nam trvali 15-20s.  
  
Vyparsovat ale potom ty data je opravdu rychla operace. Na ten nejvic komplexni soubor mi Nodejs potreboval pohych 250ms - s tim, ze jsem nedelal ani moc velke filtrace: skutecne jsem vyextrahoval vsichni bounds vsech vrstev a dokonce i dal do JSON.stringify.

Takze jedina pomalejsi operace je ten detach, ale uprimne, imo my nebudeme mit v praxi az natolik obtizne designy. Napriklad ten CBS (200mb) trva pouhych 8 sekund. Samozrejmne teoreticky muzeme narazit i na extremne komplexni designy, ale vetsinou to zakazniky delat ani nemohou z duvodu "rozbitosti" smart layoutu a obtiznosti praci s takovymi. 
  
**tl;dr:** _vetsinou zakazniky nebudou chtit velkou komplexitu layoutu, ktera je prave jedinym duvodem pomaleho detache. Detach sam o sebe je ale jedinym bottleneck mistem, takze celkove sance pomaleho zpracovani v navrzenem flow je nizka._
